### PR TITLE
Add s3 notification for lambda

### PR DESCRIPTION
**Pull request by tylersouthwick**

### CODE REVIEW

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/package.scala [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package com.monsanto.arch.cloudformation.model.resource
+
+package object resource {
+  type TopicConfigurations = NotificationConfiguration
```
Comment (tj-corrigan) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/package.scala]: Thanks for including this for backward compatibility.

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/S3.scala [hunk 2]
```diff
@@ -119,9 +119,14 @@ object S3LoggingConfiguration extends DefaultJsonProtocol {
...
 
-case class TopicConfigurations(TopicConfigurations: Seq[TopicConfiguration])
-object TopicConfigurations extends DefaultJsonProtocol {
-  implicit val format: JsonFormat[TopicConfigurations] = jsonFormat1(TopicConfigurations.apply)
+case class NotificationConfiguration(TopicConfigurations: Option[Seq[TopicConfiguration]] = None, LambdaConfigurations : Option[Seq[LambdaConfiguration]] = None)
+object NotificationConfiguration extends DefaultJsonProtocol {
+  implicit val format: JsonFormat[NotificationConfiguration] = jsonFormat2(NotificationConfiguration.apply)
+}
+
+case class LambdaConfiguration(Event: S3Event, Function: Token[String])
```
Comment (tj-corrigan) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/S3.scala]: We might as well add in Filter while we are at it.

File: src/main/scala/com/monsanto/arch/cloudformation/model/resource/S3.scala [hunk 3]
```diff
@@ -119,9 +119,14 @@ object S3LoggingConfiguration extends DefaultJsonProtocol {
   implicit val format: JsonFormat[S3LoggingConfiguration] = jsonFormat2(S3LoggingConfiguration.apply)
 }
 
-case class TopicConfigurations(TopicConfigurations: Seq[TopicConfiguration])
-object TopicConfigurations extends DefaultJsonProtocol {
-  implicit val format: JsonFormat[TopicConfigurations] = jsonFormat1(TopicConfigurations.apply)
+case class NotificationConfiguration(TopicConfigurations: Option[Seq[TopicConfiguration]] = None, LambdaConfigurations : Option[Seq[LambdaConfiguration]] = None)
```
Comment (tj-corrigan) [inline src/main/scala/com/monsanto/arch/cloudformation/model/resource/S3.scala]: Can we add QueueConfigurations  too https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket-notificationconfig-queueconfig.html

**Comment by bkrodgers** [comment]
@tylersouthwick , do you think you could add in the things TJ suggested?  If not I can merge and open another issue to add them later.

**Comment by bkrodgers** [comment]
Merging, and opened https://github.com/MonsantoCo/cloudformation-template-generator/issues/160 for the additional changes TJ suggested.

**Pull request merged**